### PR TITLE
Validate schedules

### PR DESCRIPTION
**Pull request by Jenietoc**
## What does this PR do?

In this PR we validate that the final time is never less than the initial time.

#1606 

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- [X] All possible cases were tested and all had a good response.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code and corrected any misspellings
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: components/ui/form/Schedule.tsx [hunk 1]
```diff
@@ -62,6 +65,41 @@ const TimeRangeField = ({ name }: TimeRangeFieldProps) => {
...
+        }
+      }
+      return value;
+    }
+
+    if (value.value < valueSelectStart.value) {
+      return valueSelectStart;
+    }
+    return value;
+  };
```
Comment (juanhurtado10) [inline components/ui/form/Schedule.tsx]: @Jenietoc nesting control flow structures are not a good approach. IMO, you can improve this block of code

File: components/ui/form/Schedule.tsx [hunk 2]
```diff
@@ -73,7 +111,7 @@ const TimeRangeField = ({ name }: TimeRangeFieldProps) => {
             onFocus={() => setOptions(timeOptions())}
             onBlur={() => setOptions([])}
             defaultValue={getOption(value)}
-            onChange={(option) => onChange(new Date(option?.value as number))}
+            onChange={(option) => customOnChange(onChange, option)}
```
Comment (juanhurtado10) [inline components/ui/form/Schedule.tsx]: @Jenietoc does it make sense to name a callback `customOnChange`? Will other devs understand what it does?

File: components/ui/form/Schedule.tsx [hunk 3]
```diff
@@ -62,6 +65,41 @@ const TimeRangeField = ({ name }: TimeRangeFieldProps) => {
...
   }, []);
 
+  const [valueSelectStart, setValueStart] = useState<Option>(initialValueforSelectStart);
+
+  const customOnChange = (onChange: any, option: any) => {
+    onChange(new Date(option?.value as number));
+    setValueStart(option);
+  };
+
+  const calculateRange = (value: Option) => {
```
Comment (juanhurtado10) [inline components/ui/form/Schedule.tsx]: @Jenietoc Could you please use more descriptive names in your code? Something like `calculateTimeRange`

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: components/ui/form/Schedule.tsx [hunk 4]
```diff
@@ -45,15 +45,18 @@ type TimeRangeFieldProps = {
   name: string;
 };
 
+let initialValueforSelectStart: Option;
+
+// const { i18n } = useLocale();
+const getOption = (time: ConfigType) => ({
+  value: dayjs(time).toDate().valueOf(),
+  label: dayjs(time).utc().format("HH:mm"),
+  // .toLocaleTimeString(i18n.language, { minute: "numeric", hour: "numeric" }),
```
Comment (miguelnietoa) [inline components/ui/form/Schedule.tsx]: Why don't we remove the comments?

File: components/ui/form/Schedule.tsx [hunk 5]
```diff
@@ -62,6 +65,34 @@ const TimeRangeField = ({ name }: TimeRangeFieldProps) => {
...
   }, []);
 
+  const [valueSelectStart, setValueSelectStart] = useState<Option>(initialValueforSelectStart);
+
+  const createDateAndModifyNewValue = (onChange: any, option: any) => {
+    onChange(new Date(option?.value as number));
+    setValueSelectStart(option);
+  };
+
+  const calculateTimeRange = (value: Option) => {
```
Comment (miguelnietoa) [inline components/ui/form/Schedule.tsx]: I noticed this function returns a **boolean** value, so what do you think if you renamed this by a name with prefix 'is' or 'has'? Take a look at the naming conventions for boolean functions or methods.

File: components/ui/form/Schedule.tsx [hunk 6]
```diff
@@ -62,6 +65,34 @@ const TimeRangeField = ({ name }: TimeRangeFieldProps) => {
...
+    const earlyMorningHours = [{ open: "00:00", close: "04:45" }];
+    let opened = false;
+    earlyMorningHours.forEach((item) => {
+      const open = new Date("1/1/1999 " + item.open);
+      const close = new Date("1/1/1999 " + item.close);
+      const hourDate = new Date(`1/1/1999 ${value.label}`);
+      opened = opened || (hourDate >= open && close >= hourDate);
```
Comment (miguelnietoa) [inline components/ui/form/Schedule.tsx]: I think you can refactor this `forEach` to not run through all elements but to break this once the condition `(hourDate >= open && close >= hourDate)` is true.

**Review by miguelnietoa — CHANGES_REQUESTED** [review]
This works good but some comments were added. Tell me what is your opinion about the requested changes. Good job!

File: components/ui/form/Schedule.tsx (hunk 4 above)
Comment (Jenietoc) [reply components/ui/form/Schedule.tsx]: Perfect, i will delete the comments. you are right, there is no point in having them.

File: components/ui/form/Schedule.tsx [hunk 7]
```diff
@@ -62,6 +63,36 @@ const TimeRangeField = ({ name }: TimeRangeFieldProps) => {
...
+  const isInTimeRange = (value: Option) => {
+    const earlyMorningHours = [{ open: "00:00", close: "04:45" }];
+
+    earlyMorningHours.forEach((item) => {
+      const open = new Date("1/1/1999 " + item.open);
+      const close = new Date("1/1/1999 " + item.close);
+      const hourDate = new Date(`1/1/1999 ${value.label}`);
+      if (hourDate >= open && close >= hourDate) {
+        return true;
+      }
```
Comment (miguelnietoa) [inline components/ui/form/Schedule.tsx:82]: Returning from the `forEach` callback does nothing to the outer scope. You can refactor it using a classic for, or check the `some` method!

**Review by miguelnietoa — CHANGES_REQUESTED** [review]

**Pull request merged**